File: finding-seams/python/src/room_availability/room_registry.py
from __future__ import annotations

import re
from datetime import datetime, timezone
from typing import TypedDict


class Room(TypedDict):
    id: str
    name: str


class TimeSlot(TypedDict):
    startingTime: str
    duration: int


class RoomRegistry:
    """
    Singleton-style global registry with shared mutable state.
    In real life, this would store the data somewhere using a Repository or such.
    """

    def __init__(self) -> None:
        self._rooms: dict[str, Room] = {}
        # dicts are not hashable in sets like JS object identity; list mirrors reservation history.
        self._reservations: dict[str, list[TimeSlot]] = {}

    def register(self, room: Room | None) -> None:
        if not room or not room.get("id"):
            raise ValueError("Room must have an id")
        self._rooms[room["id"]] = room

    def reserve(self, room_id: str, time_slot: TimeSlot) -> None:
        if room_id not in self._rooms:
            raise ValueError(f"Room not found: {room_id}")
        existing = self._reservations.setdefault(room_id, [])
        existing.append(time_slot)

    def is_available(self, room_id: str, time_slot: TimeSlot) -> bool:
        room_exists = self._rooms.get(room_id)
        if not room_exists:
            return False
        reservations = self._reservations.get(room_id) or []
        return not any(
            self._is_time_slots_overlapping(time_slot, reservation)
            for reservation in reservations
        )

    _STARTING_TIME_RE = re.compile(
        r"^(\d{4}-\d{2}-\d{2})T(\d{2}):(\d{2}) (CET|CEST|UTC)$"
    )

    def _parse_starting_time_ms(self, starting_time: str) -> int:
        m = self._STARTING_TIME_RE.match(str(starting_time))
        if not m:
            raise ValueError(f"Invalid startingTime: {starting_time}")
        ymd, hh, mm, tz = m.group(1), m.group(2), m.group(3), m.group(4)
        offset = "+01:00" if tz == "CET" else "+02:00" if tz == "CEST" else "Z"
        dt = datetime.fromisoformat(f"{ymd}T{hh}:{mm}:00{offset}")
        ms = int(dt.timestamp() * 1000)
        return ms

    def _slot_interval_ms(self, slot: TimeSlot) -> tuple[int, int]:
        start_ms = self._parse_starting_time_ms(slot["startingTime"])
        duration_ms = slot["duration"] * 1000 * 60
        end_ms = start_ms + duration_ms
        return start_ms, end_ms

    def _is_time_slots_overlapping(self, a: TimeSlot, b: TimeSlot) -> bool:
        a0, a1 = self._slot_interval_ms(a)
        b0, b1 = self._slot_interval_ms(b)
        return not (a0 >= b1 or a1 <= b0)


room_registry = RoomRegistry()

File: finding-seams/typescript/src/room-availability/room-registry.ts
export type TimeSlot = {
  startingTime: string
  duration: number
}

export type Room = {
  id: string
  name: string
}

// Singleton-style global registry with shared mutable state.
// In real life, this would store the data somewhere using a Repository or such.
export class RoomRegistry {
  #rooms: Map<string, Room>
  #reservations: Map<string, Set<TimeSlot>>

  constructor() {
    this.#rooms = new Map()
    this.#reservations = new Map()
  }

  // { id: string, name: string }
  register(room: Room): void {
    if (!room || !room.id) {
      throw new Error('Room must have an id')
    }

    this.#rooms.set(room.id, room)
  }

  reserve(roomId: string, timeSlot: TimeSlot): void {
    if (!this.#rooms.has(roomId)) {
      throw new Error(`Room not found: ${roomId}`)
    }

    const existing = this.#reservations.get(roomId) ?? new Set<TimeSlot>()
    existing.add(timeSlot)
    this.#reservations.set(roomId, existing)
  }

  // timeslot { startingTime: Datetime, duration: number }
  isAvailable(roomId: string, timeSlot: TimeSlot): boolean {
    const roomExists = this.#rooms.get(roomId)
    if (!roomExists) {
      return false // We don't have such rooms
    }
    const reservationsByRoomId = Array.from(this.#reservations.get(roomId) ?? [])
    return !reservationsByRoomId.some(reservation =>
      this.#isTimeSlotsOverlapping(timeSlot, reservation),
    )
  }

  // startingTime format: '2026-01-11T10:00 CET' (date, then HH:MM, then space and zone)
  #parseStartingTimeMs(startingTime: string): number {
    const m = String(startingTime).match(
      /^(\d{4}-\d{2}-\d{2})T(\d{2}):(\d{2}) (CET|CEST|UTC)$/,
    )
    if (!m) {
      throw new Error(`Invalid startingTime: ${startingTime}`)
    }
    const [, ymd, hh, mm, tz] = m
    const offset = tz === 'CET' ? '+01:00' : tz === 'CEST' ? '+02:00' : 'Z'
    const ms = Date.parse(`${ymd}T${hh}:${mm}:00${offset}`)
    if (Number.isNaN(ms)) {
      throw new Error(`Invalid startingTime: ${startingTime}`)
    }
    return ms
  }

  #slotIntervalMs(slot: TimeSlot): { startMs: number, endMs: number } {
    const startMs = this.#parseStartingTimeMs(slot.startingTime)
    const durationMs = slot.duration * 1000 * 60
    const endMs = startMs + durationMs
    return { startMs, endMs }
  }

  // Treats each slot as [start, end) in milliseconds; overlaps if ranges intersect.
  #isTimeSlotsOverlapping(a: TimeSlot, b: TimeSlot): boolean {
    const { startMs: a0, endMs: a1 } = this.#slotIntervalMs(a)
    const { startMs: b0, endMs: b1 } = this.#slotIntervalMs(b)
    return !(a0 >= b1 || a1 <= b0)
  }
}

export const roomRegistry = new RoomRegistry()
